### PR TITLE
[REEF-261]: Make the GroupCommunicationOperator layer memory efficient by introducing StreamingCodec

### PR DESCRIPTION
**Pull request by dkm2110**
[REEF-261]: Make the GroupCommunicationOperator layer memory efficient by introducing StreamingCodec

This addressed the issue by
  * allowing users directly to write streaming codecs.
  * allowing users directly to write non-streaming codecs. In this case we write the wrapper code that directly converts this codec in to streamingcodec

JIRA: [REEF-261](https://issues.apache.org/jira/browse/REEF-261)

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
+﻿using System;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs]: Add license header.

File: lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+using Org.Apache.REEF.Wake.Remote;
+using Org.Apache.REEF.Wake.Remote.Impl;
+
+namespace Org.Apache.REEF.Network.Tests.GroupCommunication
+{
+    /// <summary>
+    /// Defines streaming codec tests
+    /// </summary>
+    [TestClass]
+    public class StreamingCodecTests
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs]: Can you please add edge case tests for {{null}} passed as either data or stream?

File: lang/cs/Org.Apache.REEF.Network/Group/Config/StreamingCodecConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+using Org.Apache.REEF.Wake.Remote;
+
+namespace Org.Apache.REEF.Network.Group.Config
+{
+    /// <summary>
+    /// Defines configuration for streaming codecs and pipelie message 
+    /// streaming codecs by taking streaming codec as input.
+    /// </summary>
+    /// <typeparam name="T">Generic type of message</typeparam>
+    public sealed class StreamingCodecConfiguration<T> : ConfigurationModuleBuilder
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/StreamingCodecConfiguration.cs]: This only makes sense in the context of the group communications, right? And there, this one binding could be part of the overall configuration module. Or am I missing something?

**Comment by markusweimer** [comment]
This overall looks good. Left two small comments.

File: lang/cs/Org.Apache.REEF.Network/Group/Config/StreamingCodecConfiguration.cs (hunk 3 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/StreamingCodecConfiguration.cs]: yes you are right

File: lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs (hunk 2 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs]: What do you exactly mean? you mean in DataReader Writer etc throw an exception if null stream is passed and similarly in Codecs throw exception if data passed is null?

File: lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        [TestMethod]
+        public async Task TestCommonStreamingCodecs()
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs]: why use asyn Task? Do you control the call to those test method somewhere?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs]: I wanted to check the Write and Read async calls. Hence I need to include async in the function definition if we want to use them in the code,

File: lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/FloatStreamingCodec.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            return await reader.ReadFloatAsync(token);
+        }
+
+        /// <summary>
+        /// Writes the float to the writer.
+        /// </summary>
+        /// <param name="obj">The float to be encoded</param>
+        /// <param name="writer">The writer to which to write</param>
+        /// <param name="token">Cancellation token</param>
+        public async System.Threading.Tasks.Task WriteAsync(float obj, IDataWriter writer, CancellationToken token)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/FloatStreamingCodec.cs]: System.Threading.Tasks. is not necessary

File: lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/FloatArrayStreamingCodec.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            return floatArr;
+        }
+
+        /// <summary>
+        /// Writes the float array to the writer.
+        /// </summary>
+        /// <param name="obj">The float array to be encoded</param>
+        /// <param name="writer">The writer to which to write</param>
+        /// <param name="token">Cancellation token</param>
+        public async System.Threading.Tasks.Task WriteAsync(float[] obj, IDataWriter writer, CancellationToken token)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/FloatArrayStreamingCodec.cs]: System.Threading.Tasks. is not necessary.

File: lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/DoubleArrayStreamingCodec.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+ */
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Org.Apache.REEF.Network.StreamingCodec;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Wake.Remote;
+
+namespace Org.Apache.REEF.Network.CommonStreamingCodecs
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/DoubleArrayStreamingCodec.cs]: Why not put those classes into StreamingCodec folder?

File: lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/DoubleStreamingCodec.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            return await reader.ReadDoubleAsync(token);
+        }
+
+        /// <summary>
+        /// Writes the double to the writer.
+        /// </summary>
+        /// <param name="obj">The double to be encoded</param>
+        /// <param name="writer">The writer to which to write</param>
+        /// <param name="token">Cancellation token</param>
+        public async System.Threading.Tasks.Task WriteAsync(double obj, IDataWriter writer, CancellationToken token)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/DoubleStreamingCodec.cs]: System.Threading.Tasks. can be removed.

File: lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/FloatArrayStreamingCodec.cs (hunk 6 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/FloatArrayStreamingCodec.cs]: Removed all such occurences

File: lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/DoubleArrayStreamingCodec.cs (hunk 7 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/DoubleArrayStreamingCodec.cs]: Makes sense...I will put the folder as StreamingCodec'subfolder

File: lang/cs/Org.Apache.REEF.Network/Group/Pipelining/StreamingPipelineMessageCodec.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+ * under the License.
+ */
+
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Wake.Remote;
+using System.Threading;
+using System.Threading.Tasks;
+using Org.Apache.REEF.Network.StreamingCodec;
+
+namespace Org.Apache.REEF.Network.Group.Pipelining
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/StreamingPipelineMessageCodec.cs]: Should move this class into Group.Codec? Similar to PipelineMessageCodec. May be we don't really need Pipeline folder.

File: lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs]: Yes, those are the edge cases for which nice error messages in the exceptions are awesome :-)

File: lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs (hunk 4 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/StreamingCodecTests.cs]: Agreed. Async support for this is nice to have.

File: lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/DoubleArrayStreamingCodec.cs (hunk 7 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/CommonStreamingCodecs/DoubleArrayStreamingCodec.cs]: Also, use that namespace as well :-)

File: lang/cs/Org.Apache.REEF.Network/Group/Config/StreamingCodecConfiguration.cs (hunk 3 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/StreamingCodecConfiguration.cs]: Shall we remove it from this PR then?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/StreamingCodecConfiguration.cs]: ok

File: lang/cs/Org.Apache.REEF.Network/StreamingCodec/IStreamingCodec.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/StreamingCodec/IStreamingCodec.cs]: License header

File: lang/cs/Org.Apache.REEF.Network/StreamingCodec/CodecToStreamingCodec.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/StreamingCodec/CodecToStreamingCodec.cs]: License header

**Comment by dkm2110** [comment]
did all the required changes...please merge

**Comment by markusweimer** [comment]
This looks good to me, with one minor glitch: The changed files have lots of unused `using` statements. Can you go through them and remove?

**Comment by dkm2110** [comment]
I have fixed the using statements

File: lang/cs/Org.Apache.REEF.Network/Group/Pipelining/StreamingPipelineMessageCodec.cs (hunk 9 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/StreamingPipelineMessageCodec.cs]: We have codec folder in multiple places. Can we discuss about it?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/StreamingPipelineMessageCodec.cs]: Sure. We can file a separate jira and talk about it. Lets merge this pull
request quickly so that I can proceed to the final one :) .

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/StreamingPipelineMessageCodec.cs]: I'm not that concerned about having `ICodec` and `IStreamingCodec` *implementations* all over the place, as long as we keep the *interfaces* somewhere central. I think we have the interfaces reasonably central: `ICodec` in Wake and `IStreamingCodec` in Network. 

@jwang98052 are you ok with merging this as-is?

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/StreamingPipelineMessageCodec.cs]: @jwang98052 and I just spoke offline and decided to move ahead. I will merge this.
